### PR TITLE
Add PPS ALCA producers for a replay on Run-3 data

### PR DESCRIPTION
**Pull request by tvami**
# Replay Request

**Requestor**  

ALCADB

**Describe the configuration**  
* Release: ~CMSSW_12_5_0_pre2~, then CMSSW_12_3_6
* Run: 352929,354332
* GTs:
   * expressGlobalTag: ~123X_dataRun3_Express_v8~, then 123X_dataRun3_Express_v10
   * promptrecoGlobalTag: ~123X_dataRun3_Prompt_v10~, then 123X_dataRun3_Prompt_v12
   * alcap0GlobalTag: ~123X_dataRun3_Prompt_v10~, then 123X_dataRun3_Prompt_v12
* Additional changes:
Add new express workflow with PPS AlCa producers

**Purpose of the test**  

After CMSSW_12_5_0_pre2 comes out (later today), we'll be able to test the new PPS AlCa producers with the PPS data collected last week (cmsTalk to keep streamer files for the run is in [1]). This is a lightweight replay where all the other streams are ignored.

[1] https://cms-talk.web.cern.ch/t/streamer-files-to-be-kept-for-pps-test-run-352929/11213

**T0 Operations cmsTalk thread**  

https://cms-talk.web.cern.ch/t/replay-for-testing-pps-pcl-on-run-3-data/11398

### CODE REVIEW

**Comment by tvami** [comment]
atn: @vavati @francescobrivio @malbouis

File: etc/ReplayOfflineConfiguration.py [hunk 1]
```diff
@@ -280,6 +280,30 @@
                  maxMemoryperCore=2000,
                  dataset_lifetime=14*24*3600,#lifetime for container rules. Default 14 days
                  versionOverride=expressVersionOverride)
+                 
+addExpressConfig(tier0Config, "ALCAPPS",
+                 scenario=ppScenario,
+                 diskNode=["T2_CH_CERN"],
+                 data_tiers=["FEVT"],
+                 write_dqm=True,
+                 alca_producers=["PPSCalTrackBasedSel", "PromptCalibProdPPSTimingCalib", "PromptCalibProdPPSAlignment"],
```
Comment (francescobrivio) [inline etc/ReplayOfflineConfiguration.py:289]: I might remember wrong but isn't the alcareco producer called `PPSCalMaxTracks`?

**Review by francescobrivio — COMMENTED** [review]
Should we also add a producer (`PPSCalMaxTracks`) to the PPS dataset:
```
# PPS 2022
DATASETS = ["AlCaPPS"]
for dataset in DATASETS:
    addDataset(tier0Config, dataset,
               do_reco=False,
               scenario=ppScenario)
```
?

File: etc/ReplayOfflineConfiguration.py (hunk 1 above)
Comment (tvami) [reply etc/ReplayOfflineConfiguration.py:289]: Wasnt that for 2018 only? Maybe I remember wrong :/

Comment (vavati) [reply etc/ReplayOfflineConfiguration.py]: right. thanks

File: etc/ReplayOfflineConfiguration.py [hunk 2]
```diff
@@ -1319,7 +1343,8 @@
 
 for dataset in DATASETS:
     addDataset(tier0Config, dataset,
-               do_reco=False,
+               do_reco=True,
+               alca_producers=["PPSCalTrackBasedSel", "PromptCalibProdPPSTimingCalib", "PromptCalibProdPPSAlignment"],
```
Comment (francescobrivio) [inline etc/ReplayOfflineConfiguration.py:1347]: This is analyzed at prompt-reco level right? so the `PromptCalibProd*` producer (which are aiming at PCL) don't make much sense here. They only work at express-reco level. Did I misunderstood?

Comment (tvami) [reply etc/ReplayOfflineConfiguration.py]: Fixed!

**Comment by tvami** [comment]
test syntax please

**Comment by tvami** [comment]
> Add ctpps DQM sequance to PPS prompt

I added it here https://github.com/dmwm/T0/pull/4686/commits/b5fdea3524ed871e508127dc14c1bcbce8098b8d let me know if this was wrong of me to do.

**Comment by tvami** [comment]
test syntax please

**Comment by jhonatanamado** [comment]
@tvami . A replay node is available for this test. Just triggered yourself as soon as the CMSSW release is available.

**Comment by tvami** [comment]
Thanks @jhonatanamado great to hear that! Indeed we'll need to wait for @qliphy to upload the the new release to `cvmfs`

**Comment by vavati** [comment]
> > Add ctpps DQM sequance to PPS prompt
> 
> I added it here [b5fdea3](https://github.com/dmwm/T0/commit/b5fdea3524ed871e508127dc14c1bcbce8098b8d) let me know if this was wrong of me to do.
It should be ok. Let's see....

**Comment by francescobrivio** [comment]
run replay please

 - release is available: `/cvmfs/cms.cern.ch/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_0_pre2`

**Comment by germanfgv** [comment]
There was a problem with the definition of ALCAPPS. I'll resubmit.

**Comment by germanfgv** [comment]
run replay please

**Comment by francescobrivio** [comment]
@germanfgv could you explain the change in [c3c7586](https://github.com/dmwm/T0/pull/4686/commits/c3c75861def494cb74a2466aefe03a7ce110ae0b)? I'm a bit lost...

Also, thanks for restarting the replay! :)

**Comment by tvami** [comment]
Here is the Grafana link for the replay:
https://monit-grafana.cern.ch/d/t_jr45h7k/cms-tier0-replayid-monitoring?orgId=11&var-Bin=5m&var-ReplayID=220609100020&var-JobType=All&var-WorkflowType=All&refresh=1m

@germanfgv do I understand that correctly that nothing is running? does that mean that there is an issue?

**Comment by germanfgv** [comment]
We are running a large scale replay that's taking over a lot of resources. that's why we have a lot of idle jobs. Some of the jobs have failed and are being retried, but the failure is not likely to be recoverable:
```
Fatal Exception (Exit Code: 8006)
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 352929 lumi: 3 event: 55649 stream: 4
   [1] Running path 'dqmoffline_step'
   [2] Prefetching for module TrackToTrackComparisonHists/'hltMerged2highPurityPV'
   [3] Prefetching for module BeamSpotOnlineProducer/'offlineBeamSpot'
   [4] Calling method for module ScalersRawToDigi/'scalersRawToDigi'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: FEDRawDataCollection
Looking for module label: rawDataCollector
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.
```

**Comment by tvami** [comment]
> We are running a large scale replay that's taking over a lot of resources.

You mean besides this one? Because this is supposed to be a single stream repack/reco replay, which should not take too much resources. 

OK, that msg is beyond AlCa, I'll write to the PPS experts.

**Comment by tvami** [comment]
@vavati and others, for further debugging, here is the tarball (thanks to German)
/afs/cern.ch/user/c/cmst0/public/PausedJobs/ALCAPPS/job_2/job/WMTaskSpace/cmsRun

File: etc/ReplayOfflineConfiguration.py [hunk 3]
```diff
@@ -280,6 +280,29 @@
                  maxMemoryperCore=2000,
                  dataset_lifetime=14*24*3600,#lifetime for container rules. Default 14 days
                  versionOverride=expressVersionOverride)
+                 
+addExpressConfig(tier0Config, "ALCAPPS",
+                 scenario=ppScenario,
+                 data_tiers=["FEVT"],
+                 write_dqm=True,
```
Comment (tvami) [inline etc/ReplayOfflineConfiguration.py:287]: Maybe, for the meantime, this should be `False`?

Comment (tvami) [reply etc/ReplayOfflineConfiguration.py:287]: sorry, maybe I sent this too early... this is just about having the DQMIO output, not really about doing DQM, right?

Comment (germanfgv) [reply etc/ReplayOfflineConfiguration.py:287]: if you dont care about dqm output, yes.

Comment (tvami) [reply etc/ReplayOfflineConfiguration.py:287]: it's not really about the output, but the step itself. And also I'm not sure if we really dont need it, we'll need to hear from the PPS DQM experts

**Comment by tvami** [comment]
Hi @emanueleusai let me tag you in this as well for DQM.

**Comment by tvami** [comment]
I had this idea to explicitly add the ctpps sequence to the express job, https://github.com/dmwm/T0/pull/4686/commits/b7279a5df2ef5f0f07f230f80f83269b8700cbfb not sure if this is the correct way of dealing with it... or if it will work at all (this is really beyond AlCa at this point)

**Comment by tvami** [comment]
run replay please

**Comment by tvami** [comment]
This replay should happen only after https://github.com/dmwm/T0/pull/4689 is finished

**Comment by tvami** [comment]
Grafana link: https://monit-grafana.cern.ch/d/t_jr45h7k/cms-tier0-replayid-monitoring?orgId=11&var-Bin=5m&var-ReplayID=220611013412&var-JobType=All&var-WorkflowType=All&refresh=1m

**Comment by jhonatanamado** [comment]
Hi guys. This replay shows the same type of error that was mentioned before:
```
2022-06-11 02:28:17,780:CRITICAL:CMSSW:Error running cmsRun
{'arguments': ['/bin/bash', '/srv/job/WMTaskSpace/cmsRun1/cmsRun1-main.sh', '', 'slc7_amd64_gcc10', 'scramv1', 'CMSSW', 'CMSSW_12_5_0_pre2', 'FrameworkJobReport.xml', 'cmsRun'
, 'PSet.py', '', '', '']}
CMSSW Return code: 8006

2022-06-11 02:28:17,780:CRITICAL:CMSSW:Error message: An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 352929 lumi: 32 event: 14725218 stream: 1
   [1] Running path 'dqmoffline_step'
   [2] Prefetching for module BTagAndProbe/'BTagAndProbe_0e1m'
   [3] Prefetching for module VersionedGsfElectronIdProducer/'egmGsfElectronIDsForDQM'
   [4] Prefetching for module GEDGsfElectronFinalizer/'gedGsfElectrons'
   [5] Prefetching for module GsfElectronProducer/'gedGsfElectronsTmp'
   [6] Prefetching for module PFElecTkProducer/'pfTrackElec'
   [7] Prefetching for module GsfTrackProducer/'electronGsfTracks'
   [8] Prefetching for module CkfTrackCandidateMaker/'electronCkfTrackCandidates'
   [9] Prefetching for module ElectronSeedMerger/'electronMergedSeeds'
   [10] Prefetching for module ElectronSeedProducer/'ecalDrivenElectronSeeds'
   [11] Prefetching for module PFECALSuperClusterProducer/'particleFlowSuperClusterECAL'
   [12] Prefetching for module CorrectedECALPFClusterProducer/'particleFlowClusterECAL'
   [13] Prefetching for module PFClusterProducer/'particleFlowClusterPS'
   [14] Prefetching for module PFRecHitProducer/'particleFlowRecHitPS'
   [15] Prefetching for module ESRecHitProducer/'ecalPreshowerRecHit'
   [16] Calling method for module ESRawToDigi/'ecalPreshowerDigis'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: FEDRawDataCollection
Looking for module label: rawDataCollector
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.
```
I will fail all the jobs and kill that agent in order to have an available agent for future replays.

**Comment by tvami** [comment]
Hi @emanueleusai , this replay runs on `ALCAPPS` which by construction not contains `rawDataCollector`, so in that sense the error above is not surprising. I thought doing https://github.com/dmwm/T0/pull/4686/commits/b7279a5df2ef5f0f07f230f80f83269b8700cbfb would make the T0 only do what's needed for PPS, and nothing else. What is the way to not run the full offline DQM, just the PPS part (given that the commit above didnt work)? Thanks!

**Comment by tvami** [comment]
run replay please

**Comment by tvami** [comment]
run replay please

**Comment by tvami** [comment]
Result is unchanged
```
cmsRun1
Fatal Exception (Exit Code: 8006)
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 352929 lumi: 24 event: 10971460 stream: 1
   [1] Running path 'dqmoffline_step'
   [2] Prefetching for module CTPPSCommonDQMSource/'ctppsCommonDQMSourceOffline'
   [3] Prefetching for module CTPPSLocalTrackLiteProducer/'ctppsLocalTrackLiteProducer'
   [4] Prefetching for module CTPPSDiamondLocalTrackFitter/'ctppsDiamondLocalTracks'
   [5] Prefetching for module CTPPSDiamondRecHitProducer/'ctppsDiamondRecHits'
   [6] Calling method for module TotemVFATRawToDigi/'ctppsDiamondRawToDigi'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: FEDRawDataCollection
Looking for module label: rawDataCollector
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.
```

DQM colleagues, please comment.

**Comment by grzanka** [comment]
@tvami where can I inspect the config file being executed by T0 replay ?

I would expect that the `PPSCalMaxTrack` producer would run PPS reconstruction following the AlCa naming conventions:
https://github.com/cms-sw/cmssw/blob/c850e19102dcc5fa5978de102c57e40a7410490c/Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py
- the reco modules were renamed to end with suffix `AlCaRecoProducer` 
- the input data is expected to be under tag `hltPPSCalibrationRaw`

I am a bit surprised that module `ctppsDiamondRawToDigi` is loaded, instead of `ctppsDiamondRawToDigiAlCaRecoProducer`

**Comment by tvami** [comment]
Hi @grzanka 

@germanfgv can give you the tarball.

Regarding your surprise, I think you missed the info that all of this is happening in the DQM offline:
> 1] Running path 'dqmoffline_step'

So I believe after DQM step is passed, the ALCA steps would run what your expectation is... But that's happening only after DQM is executed

**Comment by vavati** [comment]
@tvami : i don't understand either, dqmoffline_step is executed before the reconstruction?In addition: this is supposed to run at express: dqmoffline_pps is supposed to run after the prompt (where we calculate higher lever object)
It was discussed several time: ALCAPPS should be reconstructed with the new producer, produce an AOD which will be the input for PCL. If DQM is mandatory then we need to produce a module ad-hoc as the product have a different name. But this was never discussed...in any case at this level is more appropriate to have a sort of "online" dqm as at express several calibrations are missing (which is the reason why we do this step).

**Comment by tvami** [comment]
> this is supposed to run at express: dqmoffline_pps is supposed to run after the prompt (where we calculate higher lever object)

That fails too, that was commit https://github.com/dmwm/T0/pull/4686/commits/b5fdea3524ed871e508127dc14c1bcbce8098b8d

You see I tried every possible combination, and every one of them failed :D 

> If DQM is mandatory then we need to produce a module ad-hoc as the product have a different name. 

I dont know if it's mandatory (DQM needs to comment), and I don't know how to switch it off (T0 needs to comment)

> But this was never discussed...

This is beyond AlCa, the discussion about this was supposed to happen with DQM colleagues.

**Comment by tvami** [comment]
Thanks to Javi, I learned what's the way to switch off DQM, to use the `none` in the sequence: https://github.com/dmwm/T0/pull/4686/commits/e25ca6f0e5132febdc3d059354f785362769bdaf

**Comment by jhonatanamado** [comment]
This replay shows almost the same error as before.,  but this time is with ALCARECO step
```
cmsRun1
Fatal Exception (Exit Code: 8006)
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 352929 lumi: 21 event: 9299315 stream: 1
   [1] Running path 'write_StreamALCAPPS_ALCARECO_step'
   [2] Prefetching for module PoolOutputModule/'write_StreamALCAPPS_ALCARECO'
   [3] Prefetching for module CTPPSLocalTrackLiteProducer/'ctppsLocalTrackLiteProducerAlCaRecoProducer'
   [4] Prefetching for module TotemRPLocalTrackFitter/'totemRPLocalTrackFitter'
   [5] Prefetching for module TotemRPUVPatternFinder/'totemRPUVPatternFinder'
   [6] Prefetching for module TotemRPRecHitProducer/'totemRPRecHitProducer'
   [7] Prefetching for module TotemRPClusterProducer/'totemRPClusterProducer'
   [8] Calling method for module TotemVFATRawToDigi/'totemRPRawToDigi'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: FEDRawDataCollection
Looking for module label: rawDataCollector
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.
```

**Comment by grzanka** [comment]
> @vavati DQM problem solved... now we are at least dealing with an AlCa problem.

Great !

> Looking here https://github.com/cms-sw/cmssw/blob/master/EventFilter/CTPPSRawToDigi/python/ctppsRawToDigi_cff.py#L53 and https://github.com/cms-sw/cmssw/blob/master/EventFilter/CTPPSRawToDigi/python/ctppsRawToDigi_cff.py#L101 and https://github.com/cms-sw/cmssw/blob/master/EventFilter/CTPPSRawToDigi/python/ctppsRawToDigi_cff.py#L135 they expect `rawDataCollector` while the ALCARECO expects `hltPPSCalibrationRaw` https://github.com/cms-sw/cmssw/blob/master/Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py#L17-L19
> 
> werent these supposed to have the same input? 

Are you wondering why our default reconstruction expects `rawDataCollector` rather than `hltPPSCalibrationRaw`

> I understand the second one `hltPPSCalibrationRaw` is the one that's present in the stream, correct?

That is correct, its one of the reasons of extensive renaming of modules in `PPSCalMaxTracks` producer
https://github.com/cms-sw/cmssw/blob/c850e19102dcc5fa5978de102c57e40a7410490c/Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py

**Comment by tvami** [comment]
> Are you wondering why our default reconstruction expects rawDataCollector rather than hltPPSCalibrationRaw ?

Yes indeed

**Comment by grzanka** [comment]
> > Are you wondering why our default reconstruction expects rawDataCollector rather than hltPPSCalibrationRaw ?
> 
> Yes indeed that's what I'm wondering about

In fact it was using `rawDataCollector` at least for past 6 years (looking at the commits in the repo). 
The `hltPPSCalibrationRaw` appeared when we started to introduce the new AlCa reco producer in 2021.

**Comment by tvami** [comment]
We can try to explicitly not run the reco: https://github.com/dmwm/T0/pull/4686/commits/896414417b894c794634337bc954392390701f45

**Comment by grzanka** [comment]
> This replay shows almost the same error as before., but this time is with ALCARECO step
> 
> ```
> cmsRun1
> Fatal Exception (Exit Code: 8006)
> An exception of category 'ProductNotFound' occurred while
>    [0] Processing  Event run: 352929 lumi: 21 event: 9299315 stream: 1
>    [1] Running path 'write_StreamALCAPPS_ALCARECO_step'
>    [2] Prefetching for module PoolOutputModule/'write_StreamALCAPPS_ALCARECO'
>    [3] Prefetching for module CTPPSLocalTrackLiteProducer/'ctppsLocalTrackLiteProducerAlCaRecoProducer'
>    [4] Prefetching for module TotemRPLocalTrackFitter/'totemRPLocalTrackFitter'
>    [5] Prefetching for module TotemRPUVPatternFinder/'totemRPUVPatternFinder'
>    [6] Prefetching for module TotemRPRecHitProducer/'totemRPRecHitProducer'
>    [7] Prefetching for module TotemRPClusterProducer/'totemRPClusterProducer'
>    [8] Calling method for module TotemVFATRawToDigi/'totemRPRawToDigi'

I suspect the problem may be related to the fact that PPS Silicon Strips were not included in the `PPSCalMaxTracks` AlCa reco producer (corresponding module labels were not provided with `AlCaRecoProducer` suffix and the input label was not adjusted properly).

This is however strange as we exclude strips from the reconstruction:
https://github.com/cms-sw/cmssw/blob/master/Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py#L80

I need to investigate if further by analysing provided log files

**Comment by tvami** [comment]
https://monit-grafana.cern.ch/d/t_jr45h7k/cms-tier0-replayid-monitoring?orgId=11&var-Bin=5m&var-ReplayID=220629094933&var-JobType=All&var-WorkflowType=All

I think this finished successfully , right @germanfgv  ?

**Comment by tvami** [comment]
The DAS output:
https://cmsweb.cern.ch/das/request?view=list&limit=50&instance=prod%2Fglobal&input=dataset+dataset%3D%2F*%2F*REPLAY*444%2F*+

**Comment by jhonatanamado** [comment]
Monitoring for this configuration is https://monit-grafana.cern.ch/d/t_jr45h7k/cms-tier0-replayid-monitoring?orgId=11&var-Bin=5m&var-ReplayID=220706190228&var-JobType=All&var-WorkflowType=All. This configuration shows paused jobs with the following error:
```
cmsRun1
Fatal Exception (Exit Code: 8009)
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=PoolOutputModule label='ALCARECOStreamPromptCalibProdPPSDiamondSampic'
Exception Message:
EventSelector::init, An OutputModule is using SelectEvents
to request a trigger name that does not exist
The unknown trigger name is: pathALCARECOPromptCalibProdPPSDiamondSampicTiming
```
Logs and tarballs can be found here `/afs/cern.ch/user/c/cmst0/public/Tarballs_Replays/PR4686/job_274`

**Comment by tvami** [comment]
@vavati please note the error with the SAMPIC above
@jhonatanamado I'll remove the SAMPIC now from the config, so let's do another test in a minute

**Comment by tvami** [comment]
https://monit-grafana.cern.ch/d/t_jr45h7k/cms-tier0-replayid-monitoring?orgId=11&var-Bin=5m&var-ReplayID=220707145856&var-JobType=All&var-WorkflowType=All

I think this is done

**Comment by germanfgv** [comment]
Yes @tvami. It was successful, I'll include this in the 12_3_7 replay.

**Comment by tvami** [comment]
Closing this for https://github.com/dmwm/T0/pull/4708

**Pull request closed without merging**